### ISSUE TITLE
Basic UI: Recover from connection issues

### ISSUE THREAD

**Issue by wborn** [eclipse/smarthome#3031]
The items in the Basic UI are no longer updated when the SSE stream stops working due to a connection timeout. I've looked at the Basic UI code but it seems no error handling function is assigned to the EventSource `onerror` property.

There could be several reasons for a connection timeout (or other connection related issues):
* reverse proxy or firewall connection timeout settings
* unstable WiFi or cellular connections
* unstable/restarted network equipment (switch/router)
* suspended network connections due to battery saving

Reestablishing the event stream will not be enough because events might have been missed. So perhaps the page could reload itself when the connection can be reestablished?

**Comment by maggu2810** [comment]
Re-establishing the event stream sounds good to me.
A reload makes perhaps sense, too.
IIRC the current implementation in the PaperUI implements only a re-establishing without a reload.

**Comment by kaikreuzer** [comment]
@resetnow Could you have a look at that?

**Comment by wborn** [comment]
Thanks for having a look at this issue @resetnow . The offline notification is also helpful. I just gave it a test with enabling/disabling airplane mode on my mobile phone and all works as expected. :-)

**Comment by lolodomo** [comment]
That is not clear to me. Was it fixed or not ?

**Comment by resetnow** [comment]
Sort of, it doesn't work for Chrome on Linux because the SSE disconnection event is never triggered. You can try unplugging the cable and see what happens.

**Comment by resetnow** [comment]
Just tried again with Chrome 58 on Linux — still doesn't work. Works with Firefox.

In Chrome dev tools, SSE connection is still listed as active even after the cable is unplugged — so this might be a browser issue.

**Comment by wborn** [comment]
@resetnow Did you also have a look at reconnection when the runtime is restarted? 

In case of Chrome on Linux I do see the Offline notification popup when the runtime is stopped. When I then restart the runtime, the notification is removed. But it starts to continuously do "subscribe" calls which end up in a pending state.

See this screenshot:
![restart](https://cloud.githubusercontent.com/assets/12213581/25971755/b277f140-369d-11e7-89cc-1798df765f4f.png)

**Comment by resetnow** [comment]
Stopping runtime also works for me. Didn't check the repeated calls though — it's supposed to stop subscribing once one of the calls succeeds. Not sure why they're pending.

**Comment by wborn** [comment]
@resetnow After some more testing it looks like it opens so many connections that it exhausts the connection pool of the runtime. :-) Opening any other web page served by the runtime from a second tab also fails unless the tab that exausts all the connections is closed.

**Comment by wborn** [comment]
Actually it seems to be the maximum number of connections per website that is exhausted. Accessing the page from another browser/device does seem to work.

**Comment by wborn** [comment]
When a reconnect fails it starts a new timer for each failure. Reconnection still works when I change the code to read like:

```
_t.reconnectInterval = setTimeout(function() {
```

Was there a reason for using `setInterval` here @resetnow? If you are OK with the change I'll submit a PR for it. It solves all the duplicate calls and after reconnection I'm again able to use the Basic UI without having to close the tab. :-)

**Comment by resetnow** [comment]
>setInterval

reconnection attempt may fail multiple times. Using setInterval ensures connection will be restored even if several attempts failed.

**Comment by resetnow** [comment]
I just tried restarting runtime with BasicUI page opened and I believe the issue is not related to usage of setIterval. Here is a screenshot from Firefox dev tools:

![screenshot from 2017-05-14 16-18-04](https://cloud.githubusercontent.com/assets/2057191/26034169/1c895658-38c1-11e7-91de-c1aba065ada9.png)

As you can see, after subscription servlet has started and subscribe call returned 200, BasicUI attempted to reload the current page data. However, server side component wasn't ready yet, so navigation failed and UI started trying to subscribe again.

Now, what would be the proper approach here? We could try to handle navigation errors, I guess, because server startup can take any time.

**Comment by wborn** [comment]
@resetnow  I think there are two problems:
* The issue of the servlet not yet being fully started, which we could solve by making sure both calls return 200 within a certain timeframe. The order of which one is started first may be timing dependent and could change over time.
* The existing EventSource keeps trying to reconnect (every 3s) which causes onerror to be called. This causes multiple reconnection timers to be started. If you let it run some time and add some logging it shows something like the screenshot below. Perhaps we can close the existing EventSource once  .onerror is called?

![reconnect](https://cloud.githubusercontent.com/assets/12213581/26034623/b30ed456-38bf-11e7-80c2-67e8cdb21996.png)

**Comment by resetnow** [comment]
You're right about closing EventSource connection — I tested it with Firefox, I guess reconnection behaviour is different there. I assume this will fix the problem:

```diff
		_t.source.onerror = function() {
+			_t.source.close();
			_t.connectionError();
		};
```

**Comment by wborn** [comment]
Yes that seems to work well for the EventSource on (tested Chrome desktop/mobile and Firefox). :-)

**Comment by resetnow** [comment]
Thanks for confirming — I have opened a PR for this part of the issue.

**Comment by antares2001** [comment]
Hello @resetnow 

has there been an update on this? I also face on the current unstable build of OpenHAB that when it looses connection only a page reload helps (or sometimes long waiting) to be able to interact with BasicUI. Especially when using refresh options for images (for ip cameras) in my multi-level sitemap this happens very often and makes the usage rather inconvenient (having to usually reload the page after two or three interactions).

Any news on this?

**Comment by resetnow** [comment]
@antares2001 could you please provide a bit more details?

1. What device and browser are you using to access BasicUI?
2. Does your device use wireless connection?
3. Does openHAB server use wireless connection?

**Comment by ikke-t** [comment]
I have related problem, where HAProxy in front of OpenHAB doing SSL termination and auth breaks paperUI connection. Would you know why this connection does break with proxy?

I debugged it to the state that it seems paper UI sends cookie to client app to authenticate thee rest calls, is it so? However when HAprixy adds authorization header, the cookie is never seen. Thus REST calls fail, and I get this disconnect notification.

Any idea what causes it? Is it paperui not sending cookies if there is auth header, or is it HAproxy somehow dropping cookie? More here: 

https://community.openhab.org/t/openhab-in-container-behind-haproxy-offline-popups/45320

BR,
Ilkka
